Create credentialModel elements once required data is available

AparnaKarve opened the pull request:
In order to create valid `model` and `modelCopy` objects for dirty-checking to work correctly, create (and initialize) all `credentialModel` elements  once the required data (like `vm.credential_options` and `vm.credentialModel.type`) is available.

AparnaKarve commented:
@himdel Please review in the context of https://github.com/ManageIQ/manageiq-ui-classic/pull/998

himdel commented:
Aah, definitely makes sense for that controller :).

We *may* need to address this in a more generic way if this is a problem in more places (say, update the comparison to consider `null` equal to `""`, or add a `ngModel` `parser` that converts `""` to `null`), but this looks like it should definitely fix this one :). (Cc @ZitaNemeckova)

(BTW .. you may want to know about [Object.keys](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys), something like `Object.keys(vm.credential_options[vm.credentialModel.type].attributes).forEach(function(key) {` would work too.. but either is fine :).)

AparnaKarve commented:
I have replaced the lodash `.forEach` call with `Object.keys(objs).forEach` since the tone in that controller has already been set for the `Object` library. (`Object.assign` is being used in one place)
Don't want to introduce a new library if I can avoid that.
Good call. Thanks.

himdel commented:
Thanks, yeah, makes sense, right? :)

Just one more thing, sorry..

Your addition uses `credential_options` .. so you will need to move it into the `optionsPromise.then` block, to make sure those are always available.

AparnaKarve commented:
> Your addition uses credential_options .. so you will need to move it into the optionsPromise.then block, to make sure those are always available.

True. I have moved it inside `optionsPromise.then`
And while at it, I refactored the `forEach` and `for` -- consolidated them in one `forEach` loop.

himdel commented:
So.. @AparnaKarve are you absolutely sure `response.options` and `vm.credential_options[vm.credentialModel.type].attributes` always have exactly the same set of keys?

Because if not, the latest change will break stuff..

AparnaKarve commented:
@himdel So I compared the 2 objects `response.options` and `vm.credential_options[vm.credentialModel.type]['attributes']`, and I can confirm that the `for` loop should have been based off `vm.credential_options[vm.credentialModel.type]['attributes']`.

Think of `vm.credential_options[vm.credentialModel.type]['attributes']` as the superset while `response.options` is the subset of that superset.

The API may populate the `response.options` object either partially or entirely depending on the availability of the values for that particular Credential case.

That said, I would like to anyway get a confirmation from @mzazrivec on what I said above.
The last thing I want is the form broken because of the `for`/`forEach` refactor.

mzazrivec commented:
@AparnaKarve I don't know the answer off the top of my head. This needs to be tested.

AparnaKarve commented:
@mzazrivec Yes, please. Although I have tested this myself, a second pair of eyes for testing/reviewing would be great! Thanks.

@himdel Since this basically fixes dirty-checking in the component in conjunction with #998, I would imagine we need this PR and #998, both, to be `fine/yes`.
Probably not a "blocker" material though.

himdel commented:
I don't know about `fine/yes`.. I mean, no problem adding `fine/yes` here, especially if the behaviour is identical to before, but since #998 changes how we do checkchange, I'm not quite sure about `fine` there..

ZitaNemeckova commented:
I'm not sure I want to deal with bugs it may introduce to Fine... I would leave it to Gaprindashvili as it's kinda refactoring.

AparnaKarve commented:
IMO, it should be OK to get the `form-changed` directive in Fine.
However, it should be applied only to the Ansible Credentials form, which is currently the only form that uses a Component with form fields, that way we narrow our scope for any possible bugs.

But like I said above, this is not exactly a blocker, so I'm also OK not including it in Fine, atm.

AparnaKarve commented:
Any volunteers for testing this?
@himdel @ZitaNemeckova @mzazrivec ?

AparnaKarve commented:
@himdel Based on the discussion in https://github.com/ManageIQ/manageiq/issues/14900, it looks like what we receive in `response.options` might change.
Let's hold off on testing/merging this until that gets settled. Thanks.

AparnaKarve commented:
Closing this PR, since the changes here were implemented in https://github.com/ManageIQ/manageiq-ui-classic/pull/1227.
These changes helped populate the `userid` field in the form (in https://github.com/ManageIQ/manageiq-ui-classic/pull/1227/commits/8383a4cfebbce1343142c9db945e9c2943d5f2cb), an issue that was mentioned in the blocker BZ https://bugzilla.redhat.com/show_bug.cgi?id=1446033
